# EDA

###  Dados

| Coluna | Descrição | Tipo / Unidade |
| :--- | :--- | :--- |
| **YEAR** | Ano do voo (ex.: 2015) | Inteiro |
| **MONTH** | Mês do voo (1 a 12) | Inteiro |
| **DAY** | Dia do mês do voo (1 a 31) | Inteiro |
| **DAY_OF_WEEK** | Dia da semana (1 = Segunda, 7 = Domingo) | Inteiro |
| **AIRLINE** | Código da companhia aérea (ex.: AA = American Airlines) | Categórica |
| **FLIGHT_NUMBER** | Número do voo | Inteiro |
| **TAIL_NUMBER** | Número de registro da aeronave | Texto |
| **ORIGIN_AIRPORT** | Código IATA do aeroporto de origem (ex.: ATL) | Categórica |
| **DESTINATION_AIRPORT** | Código IATA do aeroporto de destino | Categórica |
| **SCHEDULED_DEPARTURE** | Horário de partida programado | HHMM (Inteiro) |
| **DEPARTURE_TIME** | Horário real de partida | HHMM (Inteiro) |
| **DEPARTURE_DELAY** | Atraso na partida | Numérico (Minutos) |
| **TAXI_OUT** | Tempo gasto taxiando até a decolagem | Numérico (Minutos) |
| **WHEELS_OFF** | Horário em que o avião decolou | HHMM (Inteiro) |
| **SCHEDULED_TIME** | Tempo total programado de voo | Numérico (Minutos) |
| **ELAPSED_TIME** | Tempo total real de voo | Numérico (Minutos) |
| **AIR_TIME** | Tempo no ar | Numérico (Minutos) |
| **DISTANCE** | Distância entre origem e destino | Numérico (Milhas) |
| **WHEELS_ON** | Horário em que as rodas tocaram o solo | HHMM (Inteiro) |
| **TAXI_IN** | Tempo taxiando até o portão de desembarque | Numérico (Minutos) |
| **SCHEDULED_ARRIVAL** | Horário de chegada programado | HHMM (Inteiro) |
| **ARRIVAL_TIME** | Horário de chegada real | HHMM (Inteiro) |
| **ARRIVAL_DELAY** | Atraso na chegada | Numérico (Minutos) |
| **DIVERTED** | Indica se o voo foi desviado (1 = Sim, 0 = Não) | Binária |
| **CANCELLED** | Indica se o voo foi cancelado (1 = Sim, 0 = Não) | Binária |
| **CANCELLATION_REASON** | Motivo do cancelamento (A=Airline, B=Weather, C=NAS, D=Security) | Categórica |
| **AIR_SYSTEM_DELAY** | Atraso causado por controle de tráfego aéreo | Numérico (Minutos) |
| **SECURITY_DELAY** | Atraso causado por problemas de segurança | Numérico (Minutos) |
| **AIRLINE_DELAY** | Atraso causado pela companhia aérea | Numérico (Minutos) |
| **LATE_AIRCRAFT_DELAY** | Atraso causado por chegada tardia da aeronave | Numérico (Minutos) |
| **WEATHER_DELAY** | Atraso causado por condições meteorológicas | Numérico (Minutos) |

In [ ]:
#bibliotecas
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots

pd.set_option('display.max_columns', None)

: 

In [ ]:
#dados
df_flights = pd.read_csv('../data/raw/flights.csv')
df_airlines = pd.read_csv('../data/raw/airlines.csv')
df_airports = pd.read_csv('../data/raw/airports.csv')

In [ ]:
#cruzamento dos dados
df = df_flights.merge(
    df_airlines.rename(columns={'AIRLINE': 'AIRLINE_NAME'}), 
    left_on='AIRLINE', right_on='IATA_CODE', how='left'
).drop(columns=['IATA_CODE'])

df_airports = df_airports.rename(columns={'AIRPORT': 'AIRPORT_NAME'})

df = df.merge(
    df_airports.add_prefix('ORIGIN_'), 
    left_on='ORIGIN_AIRPORT', right_on='ORIGIN_IATA_CODE', how='left'
).drop(columns=['ORIGIN_IATA_CODE'])

df = df.merge(
    df_airports.add_prefix('DESTINATION_'),
    left_on='DESTINATION_AIRPORT', right_on='DESTINATION_IATA_CODE', how='left'
).drop(columns=['DESTINATION_IATA_CODE'])

df.head()

,YEAR,MONTH,DAY,DAY_OF_WEEK,AIRLINE,FLIGHT_NUMBER,TAIL_NUMBER,ORIGIN_AIRPORT,DESTINATION_AIRPORT,SCHEDULED_DEPARTURE,DEPARTURE_TIME,DEPARTURE_DELAY,TAXI_OUT,WHEELS_OFF,SCHEDULED_TIME,ELAPSED_TIME,AIR_TIME,DISTANCE,WHEELS_ON,TAXI_IN,SCHEDULED_ARRIVAL,ARRIVAL_TIME,ARRIVAL_DELAY,DIVERTED,CANCELLED,CANCELLATION_REASON,AIR_SYSTEM_DELAY,SECURITY_DELAY,AIRLINE_DELAY,LATE_AIRCRAFT_DELAY,WEATHER_DELAY,AIRLINE_NAME,ORIGIN_AIRPORT_NAME,ORIGIN_CITY,ORIGIN_STATE,ORIGIN_COUNTRY,ORIGIN_LATITUDE,ORIGIN_LONGITUDE,DESTINATION_AIRPORT_NAME,DESTINATION_CITY,DESTINATION_STATE,DESTINATION_COUNTRY,DESTINATION_LATITUDE,DESTINATION_LONGITUDE
0,2015,1,1,4,AS,98,N407AS,ANC,SEA,5,2354.0,-11.0,21.0,15.0,205.0,194.0,169.0,1448,404.0,4.0,430,408.0,-22.0,0,0,NaN,NaN,NaN,NaN,NaN,NaN,Alaska Airlines Inc.,Ted Stevens Anchorage International Airport,Anchorage,AK,USA,61.17432,-149.99619,Seattle-Tacoma International Airport,Seattle,WA,USA,47.44898,-122.30931
1,2015,1,1,4,AA,2336,N3KUAA,LAX,PBI,10,2.0,-8.0,12.0,14.0,280.0,279.0,263.0,2330,737.0,4.0,750,741.0,-9.0,0,0,NaN,NaN,NaN,NaN,NaN,NaN,American Airlines Inc.,Los Angeles International Airport,Los Angeles,CA,USA,33.94254,-118.40807,Palm Beach International Airport,West Palm Beach,FL,USA,26.68316,-80.09559
2,2015,1,1,4,US,840,N171US,SFO,CLT,20,18.0,-2.0,16.0,34.0,286.0,293.0,266.0,2296,800.0,11.0,806,811.0,5.0,0,0,NaN,NaN,NaN,NaN,NaN,NaN,US Airways Inc.,San Francisco International Airport,San Francisco,CA,USA,37.61900,-122.37484,Charlotte Douglas International Airport,Charlotte,NC,USA,35.21401,-80.94313
3,2015,1,1,4,AA,258,N3HYAA,LAX,MIA,20,15.0,-5.0,15.0,30.0,285.0,281.0,258.0,2342,748.0,8.0,805,756.0,-9.0,0,0,NaN,NaN,NaN,NaN,NaN,NaN,American Airlines Inc.,Los Angeles International Airport,Los Angeles,CA,USA,33.94254,-118.40807,Miami International Airport,Miami,FL,USA,25.79325,-80.29056
4,2015,1,1,4,AS,135,N527AS,SEA,ANC,25,24.0,-1.0,11.0,35.0,235.0,215.0,199.0,1448,254.0,5.0,320,259.0,-21.0,0,0,NaN,NaN,NaN,NaN,NaN,NaN,Alaska Airlines Inc.,Seattle-Tacoma International Airport,Seattle,WA,USA,47.44898,-122.30931,Ted Stevens Anchorage International Airport,Anchorage,AK,USA,61.17432,-149.99619


**Obtendo target**

De acordo com o Departamento de Transportes dos EUA (BTS), um voo é considerado atrasado atrasado quando chega 15 ou mais
minutos depois do horário de chegada previsto. Diante disso, vamos adotar o padrão citado como target do modelo supervisionado.

Fonte: [https://www.transtats.bts.gov/ot_delay/ot_delaycause1.asp?20=E&2r4v1qs41z=FHGDE%20&2r4v1q61=FHGDG](https://www.transtats.bts.gov/ot_delay/ot_delaycause1.asp?20=E&2r4v1qs41z=FHGDE%20&2r4v1q61=FHGDG)

In [ ]:
#obtendo variável target
df['IS_DELAYED'] = (df['ARRIVAL_DELAY'] > 15).astype(int)

**Qual é a distribuição entre voos atrasados e pontuais?**

In [ ]:
df_temp = df['IS_DELAYED'].value_counts().reset_index()
df_temp.columns = ['TARGET_STATUS', 'COUNT']
df_temp['STATUS_NAME'] = df_temp['TARGET_STATUS'].map({0: 'Pontual', 1: 'Atrasado'})

colors_map = {'Pontual': '#3498db', 'Atrasado': '#e74c3c'}
colors = [colors_map[name] for name in df_temp['STATUS_NAME']]

fig = go.Figure(data=[go.Pie(
    labels=df_temp['STATUS_NAME'],
    values=df_temp['COUNT'],
    hole=0.4,
    pull=[0, 0.1],
    marker=dict(colors=colors),
    textinfo='percent+label+value'
)])
fig.update_layout(
    title='Distribuição dos Voos',
    template='plotly_white',
    height=600
)
fig.show()

**Qual o perfil dos atrasos?**

In [ ]:
delays = ['DEPARTURE_DELAY', 'ARRIVAL_DELAY']

df_temp = df[delays].describe()
df_temp.loc['cv'] = df_temp.loc['std'] / df_temp.loc['mean']
df_temp.loc['kurtosis'] = df[delays].kurtosis()
df_temp

,DEPARTURE_DELAY,ARRIVAL_DELAY
count,5.732926e+06,5.714008e+06
mean,9.370158e+00,4.407057e+00
std,3.708094e+01,3.927130e+01
min,-8.200000e+01,-8.700000e+01
25%,-5.000000e+00,-1.300000e+01
50%,-2.000000e+00,-5.000000e+00
75%,7.000000e+00,8.000000e+00
max,1.988000e+03,1.971000e+03
cv,3.957344e+00,8.911002e+00
kurtosis,1.230056e+02,9.773980e+01


In [ ]:
labels = ['Atraso de Partida', 'Atraso de Chegada']

fig = make_subplots(
    rows=3, cols=1,
    subplot_titles=(
        'Média de Minutos', 
        'Coeficiente de Variação', 
        'Curtose'
    ),
    vertical_spacing=0.1
)
fig.add_trace(go.Bar(
    x=labels, 
    y=df_temp.loc['mean', delays],
    marker=dict(color=df_temp.loc['mean', delays], colorscale='Blues'),
    text=df_temp.loc['mean', delays].round(2),
    textposition='auto',
    name='Média'
), row=1, col=1)
fig.add_trace(go.Bar(
    x=labels, 
    y=df_temp.loc['cv', delays],
    marker=dict(color=df_temp.loc['cv', delays], colorscale='Blues'),
    text=df_temp.loc['cv', delays].round(2),
    textposition='auto',
    name='CV'
), row=2, col=1)
fig.add_hline(
    y=1, line_dash='dash', line_color='red', row=2, col=1, 
    annotation_text='Alta Instabilidade', annotation_position='top left'
)
fig.add_trace(go.Bar(
    x=labels, 
    y=df_temp.loc['kurtosis', delays],
    marker=dict(color=df_temp.loc['kurtosis', delays], colorscale='Blues'),
    text=df_temp.loc['kurtosis', delays].round(0),
    textposition='outside',
    name='Curtose'
), row=3, col=1)
fig.update_layout(
    height=1000,
    title_text='Análise de Distribuição e Volatilidade dos Atrasos',
    template='plotly_white',
    showlegend=False
)
fig.update_yaxes(title_text='Tempo (min)', row=1, col=1)
fig.update_yaxes(title_text='CV', row=2, col=1)
fig.update_yaxes(title_text='Curtose', row=3, col=1)
fig.show()

KeyError: 'mean'

Estatisticamente, a operação típica é pontual, com medianas negativas (-2 a -5 minutos) indicando que a maioria dos voos decola e pousa antes do previsto. No entanto, a média é positiva (9 minutos), sendo distorcida por atrasos catastróficos que chegam a 2.000 minutos. Essa discrepância confirma que a média aritmética não representa o comportamento real da maioria dos voos; por isso, a mediana será nossa referência para tratar dados faltantes, evitando que esses valores extremos "sujem" as informações típicas.

A instabilidade do sistema é comprovada por uma curtose altíssima (acima de 90) e um coeficiente de variação de 8.91. Na prática, isso significa que os atrasos não seguem uma distribuição "normal"; eles são eventos raros, mas quando ocorrem, são gigantescos (outliers extremos). 

Essa característica tem um impacto direto na escolha do nosso modelo: técnicas de padronização que dependem da média e do desvio padrão serão evitadas, pois seriam esmagadas por esses valores fora da curva. Em vez disso, priorizaremos modelos robustos baseados em árvores de decisão (como Random Forest e XGBoost), que são naturalmente resilientes a outliers e conseguem mapear padrões complexos de atraso sem serem enganados por casos extremos.

**Quais são as principais causas de atraso?**

In [ ]:
causes = ['AIR_SYSTEM_DELAY', 'AIRLINE_DELAY', 'LATE_AIRCRAFT_DELAY', 'SECURITY_DELAY', 'WEATHER_DELAY']

df_temp = df[causes].describe()
df_temp.loc['cv'] = df_temp.loc['std'] / df_temp.loc['mean']
df_temp.loc['kurtosis'] = df[causes].kurtosis()
df_temp

,AIR_SYSTEM_DELAY,AIRLINE_DELAY,LATE_AIRCRAFT_DELAY,SECURITY_DELAY,WEATHER_DELAY
count,1.063439e+06,1.063439e+06,1.063439e+06,1.063439e+06,1.063439e+06
mean,1.348057e+01,1.896955e+01,2.347284e+01,7.615387e-02,2.915290e+00
std,2.800368e+01,4.816164e+01,4.319702e+01,2.143460e+00,2.043334e+01
min,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
50%,2.000000e+00,2.000000e+00,3.000000e+00,0.000000e+00,0.000000e+00
75%,1.800000e+01,1.900000e+01,2.900000e+01,0.000000e+00,0.000000e+00
max,1.134000e+03,1.971000e+03,1.331000e+03,5.730000e+02,1.211000e+03
cv,2.077337e+00,2.538893e+00,1.840298e+00,2.814643e+01,7.009024e+00
kurtosis,7.152914e+01,1.348506e+02,3.207595e+01,1.014182e+04,4.516992e+02


In [ ]:
labels = ['Sistema Aéreo', 'Companhia Aérea', 'Aeronave Tardia', 'Segurança', 'Clima']

fig = make_subplots(
    rows=3, cols=1,
    subplot_titles=(
        'Média de Minutos por Causa', 
        'Coeficiente de Variação por Causa', 
        'Curtose por Causa'
    ),
    vertical_spacing=0.1
)
fig.add_trace(go.Bar(
    x=labels, 
    y=df_temp.loc['mean', causes],
    marker=dict(color=df_temp.loc['mean', causes], colorscale='Blues'),
    text=df_temp.loc['mean', causes].round(2),
    textposition='auto',
    name='Média'
), row=1, col=1)
fig.add_trace(go.Bar(
    x=labels, 
    y=df_temp.loc['cv', causes],
    marker=dict(color=df_temp.loc['cv', causes], colorscale='Blues'),
    text=df_temp.loc['cv', causes].round(2),
    textposition='auto',
    name='CV'
), row=2, col=1)

fig.add_hline(
    y=1, line_dash='dash', line_color='red', row=2, col=1, 
    annotation_text='Alta Instabilidade', annotation_position='top left'
)
fig.add_trace(go.Bar(
    x=labels, 
    y=df_temp.loc['kurtosis', causes],
    marker=dict(color=df_temp.loc['kurtosis', causes], colorscale='Blues'),
    text=df_temp.loc['kurtosis', causes].round(0),
    textposition='outside',
    name='Curtose'
), row=3, col=1)
fig.update_layout(
    height=1000,
    title_text='Análise de Distribuição e Volatilidade das Causas de Atraso',
    template='plotly_white',
    showlegend=False
)
fig.update_yaxes(title_text='Tempo (min)', row=1, col=1)
fig.update_yaxes(title_text='CV', row=2, col=1)
fig.update_yaxes(title_text='Curtose', row=3, col=1)
fig.show()

A análise das causas revela que o maior desafio do setor é a eficiência operacional. O predomínio de atrasos por aeronave tardia evidencia um efeito cascata: o sistema não consegue absorver um atraso inicial, replicando-o nos voos seguintes. Como os atrasos por companhia aérea também são significativos, fica claro que a maior parte do tempo perdido decorre de falhas na gestão logística e manutenção, superando fatores externos inevitáveis como clima ou segurança.

O atraso por aeronave tardia apresenta a maior média (23,47 min), seguido de perto pelos atrasos de companhia aérea (18,96 min). Juntos, eles formam o núcleo de ineficiência operacional que o modelo precisa priorizar. Em contrapartida, fatores como Clima e Segurança apresentam médias baixas (2,91 e 0,07 min, respectivamente), mas curtoses extremamente elevadas — no caso da segurança, ultrapassando a marca de 10.000. Isso indica que, embora esses eventos externos sejam raros, quando ocorrem, geram impactos desproporcionais que fogem completamente à regra geral da operação, reforçando a necessidade de algoritmos que não sejam sensíveis apenas às médias, mas que compreendam a gravidade desses desvios extremos.

**Quando ocorrem os atrasos?**

In [ ]:
df_temp = df.copy()

df_temp['SCHEDULED_DEPARTURE_HOUR'] = df_temp['SCHEDULED_DEPARTURE'] // 100
df_temp['SCHEDULED_TIME_HOUR'] = df_temp['SCHEDULED_TIME_HOUR'] = df_temp['SCHEDULED_TIME'] / 60

df_month = df_temp.groupby('MONTH')['IS_DELAYED'].mean().reset_index()
df_dow = df_temp.groupby('DAY_OF_WEEK')['IS_DELAYED'].mean().reset_index()
df_hour = df_temp.groupby('SCHEDULED_DEPARTURE_HOUR')['IS_DELAYED'].mean().reset_index()

fig = make_subplots(
    rows=3, cols=1, 
    subplot_titles=('% Atraso por Mês', '% Atraso por Dia da Semana', '% Atraso por Hora do Dia'),
    vertical_spacing=0.1
)
fig.add_trace(
    go.Bar(x=df_month['MONTH'], y=df_month['IS_DELAYED'], marker_color='skyblue', name='Mês'),
    row=1, col=1
)
fig.add_trace(
    go.Bar(x=df_dow['DAY_OF_WEEK'], y=df_dow['IS_DELAYED'], marker_color='salmon', name='Dia Semana'),
    row=2, col=1
)
fig.add_trace(
    go.Scatter(x=df_hour['SCHEDULED_DEPARTURE_HOUR'], y=df_hour['IS_DELAYED'], mode='lines+markers', name='Hora'),
    row=3, col=1
)
fig.update_layout(
    height=900, 
    title_text='Análise Temporal de Atrasos',
    showlegend=False,
    template='plotly_white'
)
fig.update_xaxes(title_text='Mês', row=1, col=1)
fig.update_xaxes(title_text='Dia da Semana (1=Seg, 7=Dom)', row=2, col=1)
fig.update_xaxes(title_text='Hora de Partida Programada', row=3, col=1)
fig.show()

A análise temporal revela que a probabilidade de atraso não é uma variável estática, sendo fortemente regida por ciclos de demanda e pela degradação da eficiência operacional ao longo do dia. Observamos uma **sazonalidade bimodal** nos meses de fevereiro e junho, períodos marcados por condições climáticas extremas (neve e tempestades de verão) e picos de férias. No nível semanal, a pressão do tráfego corporativo é evidente, com os maiores índices de atraso concentrados às segundas e quintas-feiras, enquanto o sábado apresenta a maior estabilidade do sistema.

O diagnóstico mais crítico, contudo, surge na análise horária: a taxa de atraso escala progressivamente de menos de 10% às 05:00 para mais de 25% às 20:00. Este comportamento é a prova visual do **efeito cascata (knock-on effect)**: conforme o dia avança, o sistema perde sua capacidade de absorver desvios, transformando atrasos pontuais de voos matinais em um congestionamento sistêmico no período noturno. Esses achados validam a necessidade de incluir variáveis temporais no modelo e são corroborados pelas análises do Departamento de Transportes dos EUA (BTS).

**Fonte**: [Bureau of Transportation Statistics (BTS) - Understanding Causes of Flight Delays](https://www.bts.gov/topics/airlines-and-airports/understanding-reporting-causes-flight-delays-and-cancellations)

**Nota sobre o Tratamento de Dados**: Durante a fase de limpeza, identificamos que os registros de outubro utilizam uma codificação numérica para os aeroportos incompatível com o padrão IATA adotado no restante do dataset. Para garantir a integridade referencial e a consistência das junções (*merges*), optou-se pela remoção estratégica deste mês, mantendo a robustez do modelo com os 11 meses restantes de dados consistentes.